drivers: gpio: Adding driver for gpio intel socfpga driver

ickochar opened the pull request:
- Added new gpio: gpio_intel_socfpga drivers for agilex socfpga
- Added gpio: gpio_intel_socfpga binding used in INTEL SOCFPGA AGILEX
- Adding gpio & pinmux node in INTEL SOCFPGA AGILEX dtsi file
- Added & updated codeowners for gpio driver

Signed-off-by: Chirag Kochar <chirag.kochar@intel.com>

Commit:
drivers: gpio: Adding new gpio_intel_socfpga driver.

Added new gpio: gpio_intel_socfpga drivers for agilex
and agilex5 socfpga

Signed-off-by: Chirag Kochar <chirag.kochar@intel.com>

Commit:
soc: arm64: intel_socfpga: Memory mapped pinmux register access.

Add memory mapped pinmux register access for INTEL AGILEX SOC FPGA.

Signed-off-by: Chirag Kochar <chirag.kochar@intel.com>

Commit:
dts: Added gpio_intel_socfpga binding & dtsi gpio node.

Added gpio: gpio_intel_socfpga binding used in INTEL SOCFPGA AGILEX
Adding gpio & pinmux nodes in INTEL SOCFPGA AGILEX dtsi file

Signed-off-by: Chirag Kochar <chirag.kochar@intel.com>

Commit:
samples: basic: blinky: Added agilex board overlay

Added INTEL SOCFPGA AGILEX board overlay to blinky sample application

Signed-off-by: Chirag Kochar <chirag.kochar@intel.com>

Commit:
CODEOWNERS: Added & updated codeowner for gpio

Adding code owner for intel socfpga gpio driver

Signed-off-by: Chirag Kochar <chirag.kochar@intel.com>

cfriedt commented on drivers/gpio/gpio_intel_socfpga.c:
```diff
@@ -0,0 +1,310 @@
+#define GPIO_INTEL_CHECK_RESET(inst)	GPIO_INTEL_INST_RESET_EN(inst) ||
+#define GPIO_INTEL_RESET_SUPPORT	(DT_INST_FOREACH_STATUS_OKAY(GPIO_INTEL_CHECK_RESET) 0)
+
+#if GPIO_INTEL_RESET_SUPPORT
```
Please remove the conditionals around the include

cfriedt replied:
I gathered that. Can you please remove the conditionals around the include? It's something we try to avoid.

cfriedt replied:
^^ please make the suggested changes

cfriedt commented on drivers/gpio/gpio_shell.c:
```diff
@@ -56,17 +58,33 @@ static int cmd_gpio_conf(const struct shell *sh, size_t argc, char **argv)
 	if (dev != NULL) {
-		index = (uint8_t)atoi(argv[args_indx.index]);
-		shell_print(sh, "Configuring %s pin %d",
+		cfg = (const struct gpio_driver_config *)dev->config;
+		if (cfg == NULL) {
+			shell_error(sh, "Device Configuration is NULL");
+			return -EIO;
+		}
+		if (index < find_msb_set(cfg->port_pin_mask)) {
+			shell_print(sh, "Configuring %s pin %u",
 			    argv[args_indx.port], index);
```
Is this change necessary? It seems to be reaching into the implementation rather than using the interface. Shouldn't the new driver Just Work (TM) with `gpio_pin_configure()`

ickochar replied:
same explanation for other comment

cfriedt commented on drivers/gpio/gpio_shell.c:
```diff
@@ -107,25 +138,43 @@ static int cmd_gpio_get(const struct shell *sh,
 	uint8_t index = 0U;
 	uint8_t value = 0U;
+	int ret;
 
 	if (isdigit((unsigned char)argv[args_indx.index][0]) != 0 &&
 	    isdigit((unsigned char)argv[args_indx.value][0]) != 0) {
 		index = (uint8_t)atoi(argv[args_indx.index]);
 		value = (uint8_t)atoi(argv[args_indx.value]);
 	} else {
-		shell_print(sh, "Wrong parameters for set");
+		shell_print(sh, "Wrong parameters for GPIO set");
 		return -EINVAL;
 	}
 	dev = device_get_binding(argv[args_indx.port]);
 
 	if (dev != NULL) {
-		index = (uint8_t)atoi(argv[2]);
-		shell_print(sh, "Writing to %s pin %d",
+		cfg = (const struct gpio_driver_config *)dev->config;
+		if (cfg == NULL) {
+			shell_error(sh, "Device Configuration is NULL");
+			return -EIO;
+		}
+		if (index < find_msb_set(cfg->port_pin_mask)) {
+			shell_print(sh, "Writing to %s pin %u",
 			    argv[args_indx.port], index);
-		gpio_pin_set(dev, index, value);
+		} else {
+			shell_error(sh, "Invalid GPIO pin");
+			return -EINVAL;
+		}
+		ret = gpio_pin_set(dev, index, value);
+		if (ret < 0) {
+			shell_error(sh, "Error %d while writing value", ret);
+			return -EIO;
+		}
+	} else {
+		shell_error(sh, "No such GPIO device");
+		return -ENODEV;
 	}
```
Same here - I'm curious as to why you need to reach into the implementation here and not simply use `gpio_pin_set()`

ickochar replied:
The changes are needed if a zephyr application has enabled gpio shell. If a user passes a invalid pin number which is more then actual number of pins available,  then following results is observed currently :
agilex$ gpio get gpio@10c03200 24
Reading gpio@10c03200 pin 24
Value 0 
while valid pin values are from 0 to 23 only.

Fix is added in gpio_shell.c for 3 reasons :
1. Few API we don't pass pin values directly to the driver api. For example, gpio_pin_get will get whole port value and then do mask :
  int (*port_get_raw)(const struct device *port,
			      gpio_port_value_t *value);

	  int (*port_set_masked_raw)(const struct device *port,
				     gpio_port_pins_t mask,
				     gpio_port_value_t value);
2. All the driver might need to have same functionality when shell is implemented.
3. Values passed above 32, will always give value 0 even if currently no. of pins above 32 doesn't exits.

New driver doesn't this this change for valid pins. This is just for invalid pins.

cfriedt commented on drivers/gpio/gpio_shell.c:
```diff
@@ -138,24 +187,36 @@ static int cmd_gpio_set(const struct shell *sh,
 static int cmd_gpio_blink(const struct shell *sh,
 			  size_t argc, char **argv)
 {
+	const struct gpio_driver_config *cfg;
 	const struct device *dev;
 	uint8_t index = 0U;
 	uint8_t value = 0U;
 	size_t count = 0;
 	char data;
+	int ret;
 
 	if (isdigit((unsigned char)argv[args_indx.index][0]) != 0) {
 		index = (uint8_t)atoi(argv[args_indx.index]);
 	} else {
-		shell_error(sh, "Wrong parameters for blink");
+		shell_error(sh, "Wrong parameters for GPIO blink");
 		return -EINVAL;
 	}
 	dev = device_get_binding(argv[args_indx.port]);
 
 	if (dev != NULL) {
-		index = (uint8_t)atoi(argv[2]);
-		shell_fprintf(sh, SHELL_NORMAL, "Blinking port %s index %d.", argv[1], index);
-		shell_fprintf(sh, SHELL_NORMAL, " Hit any key to exit");
+		cfg = (const struct gpio_driver_config *)dev->config;
+		if (cfg == NULL) {
+			shell_error(sh, "Device Configuration is NULL");
+			return -EIO;
+		}
+		if (index < find_msb_set(cfg->port_pin_mask)) {
+			shell_fprintf(sh, SHELL_NORMAL,
+			"Blinking port %s pin %u.", argv[1], index);
+			shell_fprintf(sh, SHELL_NORMAL, " Hit any key to exit");
+		} else {
+			shell_error(sh, "Invalid GPIO pin");
+			return -EINVAL;
+		}
```
Same here - it really seems like the changes to the GPIO shell are a workaround for something in the new driver.

ickochar replied:
same explanation for other comment

cfriedt reviewed (changes requested):
I think I would prefer to see this driver working without the changes to the GPIO shell.

ickochar commented:
> I think I would prefer to see this driver working without the changes to the GPIO shell.

@cfriedt Yes the driver works fine independent to shell chances.

cfriedt commented on drivers/gpio/gpio_intel_socfpga.c:
```diff
@@ -0,0 +1,310 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gpio_intel_socfpga);
+
+/* Check if reset property is defined */
+#define GPIO_INTEL_INST_RESET_EN(inst)	DT_NODE_HAS_PROP(DT_DRV_INST(inst), resets)
+#define GPIO_INTEL_CHECK_RESET(inst)	GPIO_INTEL_INST_RESET_EN(inst) ||
+#define GPIO_INTEL_RESET_SUPPORT	(DT_INST_FOREACH_STATUS_OKAY(GPIO_INTEL_CHECK_RESET) 0)
```
Please place instance macros near the bottom (a convention in Zephyr) along with other instance macros

cfriedt replied:
^^ Please address previous review comments.

Part of the reason that DT instance macros are typically defined at the bottom of the file are so that it's easier to determine if they are actually being used.

It seems to me that the above two macros are unused. They should be removed if that's the case.

ickochar replied:
As the GPIO_INTEL_RESET_SUPPORT macro is used in other part of the driver file, I have created this macro at the beginning of the file.

MaureenHelm replied:
I agree with @cfriedt . You shouldn't be using DT inst macros within the `struct gpio_intel_socfpga_config` or `gpio_init` definitions.

ickochar replied:
@MaureenHelm 
Same driver has to support multiple platform (i.e. agilex7 & agilex5) which is the main factor for implementation like this. Its possible that platform could have reset manager and some platform could not have reset manager support. So we have to check if the reset manager property is defined in a node. If yes, then we should toggle the reset manager pin to reset the peripheral during initialization of the driver.

cfriedt reviewed (changes requested):
Please make changes to the gpio shell in a separate PR and make the remaining minor changes requested.

ickochar commented:
Could we re-open this review ?

cfriedt commented on drivers/gpio/gpio_intel_socfpga.c:
```diff
@@ -0,0 +1,310 @@
+	const uint32_t gpio_port;
+	/** GPIO pin number per port */
+	uint32_t ngpios;
+#if WDT_DW_RESET_SUPPORT
```
This seems out of place. Is there a Kconfig value missing? I can't see this symbol being used anywhere in Zephyr currently.

ickochar replied:
Updated to correct GPIO based Macro

cfriedt commented on soc/arm64/intel_socfpga/agilex/mmu_regions.c:
```diff
@@ -26,6 +26,12 @@ static const struct arm_mmu_region mmu_regions[] = {
 			      DT_REG_SIZE(DT_NODELABEL(uart0)),
 			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_DEFAULT_SECURE_STATE),
 
+	MMU_REGION_FLAT_ENTRY("PINMUX",
```
I'm not super familiar with this subject, so perhaps @gmarull could take a look.

cfriedt reviewed (changes requested):
I would have personally left this PR closed as it grew stale since the author ignored previous reviews, had not made requested changes, and accrued merge conflicts.

Since it has been reopened though.. still the same thing. Please address review comments.

I did find additional items of concern as well, so I highlighted those too.

ickochar commented:
> I think I would prefer to see this driver working without the changes to the GPIO shell.
Separated to new PR

ickochar commented:
> I would have personally left this PR closed as it grew stale since the author ignored previous reviews, had not made requested changes, and accrued merge conflicts.
Apologies for inconvenience. I have addressed all the review comments all other in last push.

ickochar commented:
Hello All,

Could you please help to review this PR @cfriedt @carlocaione @gmarull @SgrrZhf

ickochar commented:
I would like to request to add dev-review label.

cfriedt reviewed (approved):
No particular red flags for me at this point. Sorry I have not been active as GPIO collaborator for some time.

henrikbrixandersen commented on drivers/gpio/gpio_intel_socfpga.c:
```diff
@@ -0,0 +1,305 @@
+#define GPIO_INTEL_CHECK_RESET(inst)	GPIO_INTEL_INST_RESET_EN(inst) ||
+#define GPIO_INTEL_RESET_SUPPORT	(DT_INST_FOREACH_STATUS_OKAY(GPIO_INTEL_CHECK_RESET) 0)
+
+#define DEV_CFG(_dev) ((const struct gpio_intel_socfpga_config *)(_dev)->config)
+#define DEV_DATA(_dev) ((struct gpio_intel_socfpga_data *const)(_dev)->data)
```
Please avoid macros like this. Just access the structure members directly.

ickochar replied:
We are memory mapping device register using DEVICE_MMIO_NAMED_MAP macros and this macro requires DEV_DATA and DEV_CFG macros to be defined. Please check below link :

https://github.com/zephyrproject-rtos/zephyr/blob/71fd9b2e07a6fa9d7895759fc83685b6849aebb4/include/zephyr/sys/device_mmio.h#L520

henrikbrixandersen commented on drivers/gpio/gpio_intel_socfpga.c:
```diff
@@ -0,0 +1,305 @@
+		}
+
+		/* Getting pinmux base address from pinmux node */
+		pmux_value = sys_read32((DT_REG_ADDR(DT_NODELABEL(pinmux)) + addroffst));
```
I'd avoid using a these macros deep in the code as it hides the dependency.

ickochar replied:
In understand your main concern related to complete pinmux/pinctrl implementation. 
Currently another hardware outside zephyr world, configuring pin controller register on our platform. 
Our main design goal is to make sure pinctrl is configured in GPIO mode before performing any GPIO operation. 
Also, Unfortunately, Pinmux driver does not have get pinctrl register values. If you say, I can create a RFC for pinctrl enhancement separately. Is that fine ?

henrikbrixandersen commented on drivers/gpio/gpio_intel_socfpga.c:
```diff
@@ -0,0 +1,305 @@
+					 gpio_port_pins_t mask,
+					 gpio_port_value_t value)
+{
+	return -ENOSYS;
```
Why?

henrikbrixandersen commented on drivers/gpio/gpio_intel_socfpga.c:
```diff
@@ -0,0 +1,305 @@
+		.gpio_port = _inst,							\
+		.ngpios = DT_INST_PROP(_inst, ngpios),					\
+		IF_ENABLED(DT_INST_NODE_HAS_PROP(inst, resets),				\
+			(DW_WDT_RESET_SPEC_INIT(inst)))					\
```
WDT?

henrikbrixandersen reviewed (changes requested):
There seem to be some confusion regarding the added pinmux. Should it be using the pinctrl driver API? Should it be using the syscon API? Current access to it is done directly, which hides the dependency deep with the code.

MaureenHelm commented on samples/basic/blinky/boards/intel_socfpga_agilex_socdk.overlay:
```diff
@@ -0,0 +1,24 @@
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	aliases {
+		led0 = &myled0;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+		myled0: led_0 {
+			gpios = <&gpio1 19 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&gpio1 {
+	status = "okay";
+};
```
Why not put these in the board devicetree?

ickochar replied:
We are currently try to enable as minimum as possible in board dtsi file. GPIO is not an essential peripheral for our platform bring up. So by default all non-essential peripheral are turned off. 
Also our pins could be re-configured for various other purpose like SPI, I2C, SDMMC , etc, Depending on end application use case, respective peripherals could be enabled flexibly. For example as this is blinky application, we can enable only GPIO peripheral here with required alias for sample application.

MaureenHelm commented on CODEOWNERS:
```diff
@@ -234,6 +234,7 @@
 /drivers/flash/*nrf*                      @de-nordic
 /drivers/flash/*esp32*                    @sylvioalves
 /drivers/gpio/*b91*                       @andy-liu-telink
+/drivers/gpio/*intel_socfpga*             @alifzakuan @ickochar
```
Use the MAINTAINERS file instead of CODEOWNERS
